File: frontend/src/types/telemetry.ts
import { TelemetryMetricKey } from "@/lib/telemetryMetrics";

export type TelemetryPoint = {
  timestamp: string;

  rpm: number;

  speedMph: number;

  throttlePercent: number;

  coolantTempF: number;

  intakeTempF?: number;

  boostPsi?: number;

  fuelLevelPercent?: number;

  batteryVoltage?: number;
};

export type LiveTelemetryPoint = {
    timestamp: string;
} & Partial <
    Record <
        TelemetryMetricKey,
        number | null
    >
>;

export type LiveSessionStats = {
    duration_seconds: number;
    distance_miles: number;
    max_speed_mph: number;
    avg_speed_mph: number;
    max_rpm: number;
    telemetry_count: number;
    speed_sum_mph: number;
};

File: backend/app/schemas/telemetry.py
from datetime import datetime
from pydantic import BaseModel


class TelemetryPointCreate(BaseModel):
    id: str
    sessionId: str

    timestamp: datetime | None = None

    rpm: int
    speed_mph: float
    throttle_percent: float
    coolant_temp_f: float

    intake_temp_f: float | None = None
    boost_psi: float | None = None
    fuel_level_percent: float | None = None
    battery_voltage: float | None = None


class TelemetryBatchCreate(BaseModel):
    points: list[TelemetryPointCreate]


class TelemetryPointResponse(BaseModel):
    id: str
    session_id: str
    timestamp: datetime

    rpm: int
    speed_mph: float
    throttle_percent: float
    coolant_temp_f: float

    intake_temp_f: float | None
    boost_psi: float | None
    fuel_level_percent: float | None
    battery_voltage: float | None

    class Config:
        from_attributes = True